src/docbuilder/caching.rs

```diff
@@ -0,0 +1,112 @@
+use crate::utils::copy_dir_all;
+use anyhow::{Context as _, Result};
+use std::{
+    fs, io,
+    path::{Path, PathBuf},
+};
+use tracing::{debug, instrument, warn};
+
+/// move cache folder to target, falling back to copy + delete on error.
+fn move_or_copy<P: AsRef<Path> + std::fmt::Debug, Q: AsRef<Path> + std::fmt::Debug>(
+    source: P,
+    dest: Q,
+) -> io::Result<()> {
+    if let Some(parent) = dest.as_ref().parent() {
+        fs::create_dir_all(parent)?;
+    }
+    if let Err(err) = fs::rename(&source, &dest) {
+        warn!(
+            ?err,
+            ?source,
+            ?dest,
+            "could not move target directory, fall back to copy"
+        );
+        copy_dir_all(&source, &dest)?;
+        fs::remove_dir_all(&source)?;
+    }
+    Ok(())
+}
+
+/// artifact caching with cleanup
+#[derive(Debug)]
+pub(crate) struct ArtifactCache {
+    cache_dir: PathBuf,
+}
+
+impl ArtifactCache {
+    pub(crate) fn new(cache_dir: PathBuf) -> Result<Self> {
+        Ok(Self { cache_dir })
+    }
+
+    pub(crate) fn purge(&self) -> Result<()> {
+        fs::remove_dir_all(&self.cache_dir)?;
+        Ok(())
+    }
+
+    /// clean up a target directory.
+    ///
+    /// Should delete all things that shouldn't leak between
+    /// builds, so:
+    /// - doc-output
+    /// - ...?
+    #[instrument(skip(self))]
+    fn cleanup(&self, target_dir: &Path) -> Result<()> {
+        for item in fs::read_dir(target_dir)? {
```

Comment:
does this cleanup work as it should? Or should we be more intelligent and use the metadata?

Reply:
I think this is ok. I don't have objections to using metadata, but there's some post-processing necessary - here's what cargo's output looks like:
```
 Documenting example v0.1.0 (/home/jyn/src/example)
{"reason":"compiler-artifact","package_id":"example 0.1.0 (path+file:///home/jyn/src/example)","manifest_path":"/home/jyn/src/example/Cargo.toml","target":{"kind":["lib"],"crate_types":["lib"],"name":"example","src_path":"/home/jyn/src/example/src/lib.rs","edition":"2021","doc":true,"doctest":true,"test":true},"profile":{"opt_level":"0","debuginfo":2,"debug_assertions":true,"overflow_checks":true,"test":false},"features":[],"filenames":["/home/jyn/src/example/target/doc/example/index.html"],"executable":null,"fresh":false}
{"reason":"build-finished","success":true}
```
So we'd need to go from `"/home/jyn/src/example/target/doc/example/index.html"` to `"/home/jyn/src/example/target/doc"]` and then remove everything recursively under it.